main_test.go

```diff
@@ -512,10 +515,71 @@ var testCases = []TestCase{
 	},
 }
 
+type FakeS3Client struct {
+	Get mockGetObjectAPI
+	Put mockPutObjectAPI
+}
+
+type mockGetObjectAPI func(ctx context.Context, params *s3.GetObjectInput) (*s3.GetObjectOutput, error)
+type mockPutObjectAPI func(ctx context.Context, params *s3.PutObjectInput) (*s3.PutObjectOutput, error)
+
+func (fc *FakeS3Client) GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+	return fc.Get(ctx, params)
+}
+
+func (fc *FakeS3Client) PutObject(ctx context.Context, params *s3.PutObjectInput, optFns ...func(*s3.Options)) (*s3.PutObjectOutput, error) {
+	return fc.Put(ctx, params)
+}
+
+func (fc *FakeS3Client) New(bucket, key string) *FakeS3Client {
+	return &FakeS3Client{
+		Get: mockGetObjectAPI(func(ctx context.Context, params *s3.GetObjectInput) (*s3.GetObjectOutput, error) {
+			if aws.StringValue(params.Bucket) != bucket {
+				return nil, fmt.Errorf("expected bucket %s; got %s", aws.StringValue(params.Bucket), bucket)
+			}
+			if aws.StringValue(params.Key) != key {
+				return nil, fmt.Errorf("expected key %s; got %s", aws.StringValue(params.Key), key)
+			}
+
+			f, err := os.Open(path.Join("/", key))
```

Comment:
If you are using a fixed key, I think there is no reason to actually store the file at the key; it adds an additional unnecessary constraint for the key (must work with filesystem).

How about `const fakeS3FileLocation = /tmp/s3file` and referring to that when you read and write the file?

Reply:
`main_test.go` creates a file object from the inputs and stores it at `/tmp/macfinxyz/in.xlsx` . The fake S3 client downloads the file by reading `/tmp/macfinxyz/in.xlsx`. The base filename (`in.xlsx`) is the base of the input key, not the entire key. (Is that a constraint?) Using `fakes3FileLocation` means the filename variable in each test case must also be set to this const, right?

Reply:
Oh yeah, I forgot that we use a temporary variable. So then that could be one more field on the struct:
```
type FakeS3Client struct {
	Bucket, Key string
        LocalPath string
}
```
and then it could read/write from/to `LocalPath`

The point is that there's no reason to tie the local storage path to the S3 key path when you are only dealing with a single key, and it means you need to worry about more things (like whether the S3 key makes sense as a local storage path).

Reply:
Thank you for the clarification. The point is to store the downloaded file (using real or fake s3 client) in a location that does not depend on the key value.